src/libraries/Microsoft.PowerFx.Core/Utils/DName.cs

```diff
@@ -1,179 +1,205 @@
+        {
+            Contracts.AssertValueOrNull(other);
+            return Value == other;
+        }
+
+        public static bool operator ==(DName name1, DName name2) => name1.Value == name2.Value;
+
+        public static bool operator ==(string str, DName name)
+        {
+            Contracts.AssertValueOrNull(str);
+            return str == name.Value;
+        }
+
+        public static bool operator ==(DName name, string str)
+        {
+            Contracts.AssertValueOrNull(str);
+            return name.Value == str;
+        }
+
+        public static bool operator !=(DName name1, DName name2) => name1.Value != name2.Value;
+
+        public static bool operator !=(string str, DName name)
+        {
+            Contracts.AssertValueOrNull(str);
+            return str != name.Value;
+        }
+
+        public static bool operator !=(DName name, string str)
+        {
+            Contracts.AssertValueOrNull(str);
+            return name.Value != str;
+        }
+
+        /// <summary>
+        /// Returns whether the given name is a valid <see cref="DName" />. 
+        /// </summary>
+        /// <param name="strName"></param>
+        /// <returns></returns>
+        public static bool IsValidDName(string strName)
+        {
+            Contracts.AssertValueOrNull(strName);
+
+            if (string.IsNullOrEmpty(strName))
+            {
+                return false;
+            }
+
+            for (var i = 0; i < strName.Length; i++)
+            {
+                var ch = strName[i];
+                if (!char.IsWhiteSpace(ch))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Takes a name and makes it into a valid <see cref="DName" />.
+        /// If the name contains all spaces, an underscore is prepended to the name.
+        /// </summary>
+        /// <param name="strName"></param>
+        /// <param name="fModified">Whether it had to be changed to be a valid <see cref="DName" />.</param>
+        public static DName MakeValid(string strName, out bool fModified)
+        {
+            Contracts.AssertValueOrNull(strName);
+
+            if (string.IsNullOrEmpty(strName))
+            {
+                fModified = true;
+                return new DName(StrUnderscore);
+            }
+
+            var fAllSpaces = true;
+            var fHasSpecialWhiteSpaceCharacters = false;
+            
+            fModified = false;
+
+            for (var i = 0; i < strName.Length; i++)
+            {
+                bool fIsSpace = strName[i] == ChSpace;
+                bool fIsWhiteSpace = char.IsWhiteSpace(strName[i]);
+                fAllSpaces = fAllSpaces && fIsWhiteSpace;
+                fHasSpecialWhiteSpaceCharacters = fHasSpecialWhiteSpaceCharacters || (fIsWhiteSpace && !fIsSpace);
+            }
+
+            if (fHasSpecialWhiteSpaceCharacters)

```

Comment:
>if (fHasSpecialWhiteSpaceCharacters) [](http://example.com/codeflow?start=12&length=36)

is this possible to encode an invalid dname differently than before?

Reply:
no, this only replaces some characters with spaces which are valid.